### ISSUE TITLE
Inconsistent query behaviour w/ lots of joins

### ISSUE THREAD

**Issue by SStorm** [crate/crate#13361]
### CrateDB version

5.1.1

### CrateDB setup information

CrateDB Cloud cluster with 3 nodes

### Steps to Reproduce

```sql
SELECT DISTINCT
    o.name as "Organization",
    p.region as "Region",
    c.name as "Cluster Name",
    coalesce(al2.context['current_user']['email'], al.context['current_user']['email']) as "Email",
    IF(RIGHT(al.context['current_user']['email'],8) = 'crate.io',1,0) as "Crate.io",
    sub.provider as "Billing Provider",
    stats.credits_used / 100000 / 100 as "$ Credits used",
    ((credits.amount - stats.credits_used) / 100000 / 100) as "$ Credits remaining",
    DATE_FORMAT('%Y-%m-%d %H:%i', sub.dc['created']) as "Subscription created",
    DATE_FORMAT('%Y-%m-%d %H:%i', c.dc['created']) as "Cluster created",
    'croud clusters delete --region ' || p.region || ' --cluster-id ' || c.id || ' --sudo' as "croud delete stmt",
    c.id as "Cluster ID",
    o.id as "Organization ID",
    c.project_id as "Project ID",
    c.crate_version as "CrateDB Version",
    sc.id as "Stripe Customer ID"
from core.clusters c
    JOIN core.projects p ON c.project_id = p.id
    JOIN core.organizations o ON p.organization_id = o.id
    LEFT JOIN core.subscriptions sub ON c.subscription_id = sub.id
    JOIN core.audit_logs al ON c.id = al.context['cluster']['id']
    LEFT JOIN core.audit_logs al2 ON (c.id = al2.context['cluster']['id'] and al2.action = 'product.initiate_create_cluster')
    LEFT JOIN core.stripe_customers sc ON sc.organization_id = o.id
    JOIN stats.credit_usage_by_cluster stats ON stats.cluster_id = c.id
    JOIN core.credits credits ON credits.organization_id = o.id
where c.is_deleted = false AND al.action = 'product.create_cluster' AND sub.state = 'ACTIVE'
  AND stats.last_ts > NOW() - interval '24 hours'
  AND p.region != 'bregenz.a1'
ORDER BY 7 desc, 1,
    2 limit 100;
```

### Expected Result

Query consistently returns same result when data does not change

### Actual Result

Query returns 0 results most of the time, but sometimes returns the expected results.

**Comment by matriv** [comment]
Hi @SStorm, Thx for reporting, have you tried to remove tables from the join one-by-one to try narrow down, which one could cause the issue?

**Comment by SStorm** [comment]
Hey @matriv , looks like it is the 2nd join with the `audit_logs` table:

```sql
SELECT DISTINCT
    o.name as "Organization",
    p.region as "Region",
    c.name as "Cluster Name",
    --coalesce(al2.context['current_user']['email'], al.context['current_user']['email']) as "Email",
    al.context['current_user']['email'] as "Email",
    IF(RIGHT(al.context['current_user']['email'],8) = 'crate.io',1,0) as "Crate.io",
    sub.provider as "Billing Provider",
    stats.credits_used / 100000 / 100 as "$ Credits used",
    DATE_FORMAT('%Y-%m-%d %H:%i', sub.dc['created']) as "Subscription created",
    DATE_FORMAT('%Y-%m-%d %H:%i', c.dc['created']) as "Cluster created",
    'croud clusters delete --region ' || p.region || ' --cluster-id ' || c.id || ' --sudo' as "croud delete stmt",
    c.id as "Cluster ID",
    o.id as "Organization ID",
    c.project_id as "Project ID",
    c.crate_version as "CrateDB Version",
    sc.id as "Stripe Customer ID"
from core.clusters c
    JOIN core.projects p ON c.project_id = p.id
    JOIN core.organizations o ON p.organization_id = o.id
    LEFT JOIN core.subscriptions sub ON c.subscription_id = sub.id
    JOIN core.audit_logs al ON c.id = al.context['cluster']['id']
    --LEFT JOIN core.audit_logs al2 ON (c.id = al2.context['cluster']['id'] and al2.action = 'product.initiate_create_cluster')
    LEFT JOIN core.stripe_customers sc ON sc.organization_id = o.id
    JOIN stats.credit_usage_by_cluster stats ON stats.cluster_id = c.id
where c.is_deleted = false AND al.action = 'product.create_cluster' AND sub.state = 'ACTIVE'
  AND stats.last_ts > NOW() - interval '24 hours'
  AND p.region != 'bregenz.a1'
limit 100;
```

works absolutely fine. As soon as I add it back in, it's inconsistent again.

**Comment by matriv** [comment]
great, thx!

**Comment by matriv** [comment]
The problem is fixed if we move the 2nd part of the join: `al2.action = 'product.initiate_create_cluster')` condition to the WHERE clause:
```
SELECT DISTINCT
    o.name as "Organization",
    p.region as "Region",
    c.name as "Cluster Name",
    coalesce(al2.context['current_user']['email'], al.context['current_user']['email']) as "Email",
    IF(RIGHT(al.context['current_user']['email'],8) = 'crate.io',1,0) as "Crate.io",
    sub.provider as "Billing Provider",
    stats.credits_used / 100000 / 100 as "$ Credits used",
    ((credits.amount - stats.credits_used) / 100000 / 100) as "$ Credits remaining",
    DATE_FORMAT('%Y-%m-%d %H:%i', sub.dc['created']) as "Subscription created",
    DATE_FORMAT('%Y-%m-%d %H:%i', c.dc['created']) as "Cluster created",
    'croud clusters delete --region ' || p.region || ' --cluster-id ' || c.id || ' --sudo' as "croud delete stmt",
    c.id as "Cluster ID",
    o.id as "Organization ID",
    c.project_id as "Project ID",
    c.crate_version as "CrateDB Version",
    sc.id as "Stripe Customer ID"
from core.clusters c
    JOIN core.projects p ON c.project_id = p.id
    JOIN core.organizations o ON p.organization_id = o.id
    LEFT JOIN core.subscriptions sub ON c.subscription_id = sub.id
    JOIN core.audit_logs al ON c.id = al.context['cluster']['id']
    LEFT JOIN core.audit_logs al2 ON c.id = al2.context['cluster']['id']
    LEFT JOIN core.stripe_customers sc ON sc.organization_id = o.id
    JOIN stats.credit_usage_by_cluster stats ON stats.cluster_id = c.id
    JOIN core.credits credits ON credits.organization_id = o.id
where c.is_deleted = false AND al.action = 'product.create_cluster' AND sub.state = 'ACTIVE'
  AND stats.last_ts > NOW() - interval '24 hours'
  AND p.region != 'bregenz.a1' AND al2.action = 'product.initiate_create_cluster')
ORDER BY 7 desc, 1,
    2 limit 100;
```

**Comment by matriv** [comment]
I haven't manage to reproduce the issue locally both in single node and a 3-node cluster. One example of what I've tried:
```
create table t1(a int primary key, b OBJECT); 
create table t2(a int primary key, b OBJECT); 
create table t3(a int primary key, b OBJECT); 
create table t4(a int primary key, b OBJECT); 
```

```
insert into t1(a) select generate_series(1, 10, 1); 
update t1 set b['foo'] = a * 10 + a; 

insert into t2(a) select generate_series(3, 12, 1); 
update t2 set b['foo'] = a * 10 + a; 

insert into t3(a) select generate_series(6, 14, 1); 
update t3 set b['foo'] = a * 10 + a; 

insert into t4(a) select generate_series(8, 16, 1); 
update t4 set b['foo'] = a * 10 + a; 

refresh tables t1,t2,t3,t4;
```

```
SELECT DISTINCT t1.a, t2.a, t3.a, t4.a 
FROM t1 
INNER JOIN t2 ON t1.b['foo'] = t2.b['foo']
LEFT OUTER JOIN t3 ON t1.b['foo'] = t3.b['foo'] AND t3.a < 8 
LEFT OUTER JOIN t4 ON t2.a * 10 + t4.a = t4.b['foo']; 
```

```
SELECT DISTINCT t1.a, t2.a, t3.a, t4.a 
FROM t1
INNER JOIN t2 ON t1.b['foo'] = t2.b['foo']
LEFT OUTER JOIN t3 ON t1.b['foo'] = t3.b['foo'] AND t3.a < 8
INNER JOIN t4 ON t2.a * 10 + t4.a = t4.b['foo'];
```

**Comment by seut** [comment]
@matriv

>The difference in the plan is that in the 1st case (original query) we get NestedLoopJoin because the 2nd part of the join condition doesn't allow the planner to recognize an equi-join. In the second case (2nd part is moved to WHERE) we get HashJoin

Where is the `NestedLoopJoin` turning into a `HashJoin`? A simple diff of your posted plans only highlights the expected change from `LEFT` to `INNER` and the join condition. Do I miss something or did you maybe mix something up with the posted plans? 
The most significant change I see is that in one plan the constant join condition is pushed to the collect phase (guess when manually moving it to the WHERE clause) while in the non working case it isn't pushed down.

**Comment by proddata** [comment]
```sql
SELECT 
    c.name
from core.clusters c
    JOIN core.projects p ON c.project_id = p.id
    JOIN core.organizations o ON p.organization_id = o.id
    LEFT JOIN core.audit_logs al2 ON (c.id = al2.context['cluster']['id'] and al2.action = 'product.initiate_create_cluster')
    LEFT JOIN core.stripe_customers sc ON sc.organization_id = o.id
    JOIN core.credits credits ON credits.organization_id = o.id
where c.is_deleted = false
```

is enough to sometimes get `0` results

the `context` field is of type `OBJECT(IGNORED)` which might also have an effect on the physical plan?!

```sql
CREATE TABLE  "core"."audit_logs"
(
    "id"                   TEXT NOT NULL,
    "action"               TEXT,
    "context"              OBJECT(IGNORED),
    PRIMARY KEY ("id")
) CLUSTERED BY ("id")

**Comment by matriv** [comment]
> > The difference in the plan is that in the 1st case (original query) we get NestedLoopJoin because the 2nd part of the join condition doesn't allow the planner to recognize an equi-join. In the second case (2nd part is moved to WHERE) we get HashJoin
> 
> Where is the `NestedLoopJoin` turning into a `HashJoin`? A simple diff of your posted plans only highlights the expected change from `LEFT` to `INNER` and the join condition. Do I miss something or did you maybe mix something up with the posted plans? The most significant change I see is that in one plan the constant join condition is pushed to the collect phase (guess when manually moving it to the WHERE clause) while in the non working case it isn't pushed down.

@seut You're right, apologies, I probably mixed up with some other versions of the query I was trying out, no NL<->HashJoin change, just the LEFT to INNER.

**Comment by matriv** [comment]
The conversion of LEFT JOIN to INNER when a filter (WHERE clause) on the right relation of the LEFT join is applied is valid.
from https://stackoverflow.com/questions/3256304/left-join-turns-into-inner-join:
```
Whenever you specify a value from the right side of a left join in a WHERE clause (which is NOT NULL), you necessarily eliminate all of the NULL values and it essentially becomes an INNER JOIN.

If you write, AND (c.foobar = 'somethingelse' OR c.foobar IS NULL) that will solve the problem.

You can also move the c.foobar portion into your join predicate, and that too will solve the issue.
```
This in our code is applied here: https://github.com/crate/crate/blob/master/server/src/main/java/io/crate/planner/optimizer/rule/RewriteFilterOnOuterJoinToInnerJoin.java#L163

Checkout PSQL behaviour:
```
matriv=> SELECT t1.a,t1.b,tt2.a,tt2.b 
         FROM t1  
         LEFT OUTER JOIN t2 AS tt2 ON t1.a = tt2.a;
 a  |  b  | a | b  
----+-----+---+----
  1 |  11 |   |   
  2 |  22 |   |   
  3 |  33 | 3 | 33
  4 |  44 | 4 | 44
 15 |  55 |   |   
 16 |  66 |   |   
 17 |  77 |   |   
 18 |  88 |   |   
 19 |  99 |   |   
 20 | 110 |   |   
(10 rows)

matriv=> EXPLAIN SELECT t1.a,t1.b,tt2.a,tt2.b 
         FROM t1  
         LEFT OUTER JOIN t2 AS tt2 ON t1.a = tt2.a;
                              QUERY PLAN                              
----------------------------------------------------------------------
 Hash Left Join  (cost=60.85..99.39 rows=2260 width=16)
   Hash Cond: (t1.a = tt2.a)
   ->  Seq Scan on t1  (cost=0.00..32.60 rows=2260 width=8)
   ->  Hash  (cost=32.60..32.60 rows=2260 width=8)
         ->  Seq Scan on t2 tt2  (cost=0.00..32.60 rows=2260 width=8)
(5 rows)

matriv=> SELECT t1.a,t1.b,tt2.a,tt2.b 
         FROM t1  
         LEFT OUTER JOIN t2 AS tt2 ON t1.a = tt2.a WHERE tt2.b > 20;

 a | b  | a | b  
---+----+---+----
 3 | 33 | 3 | 33
 4 | 44 | 4 | 44
(2 rows)


matriv=> EXPLAIN SELECT t1.a,t1.b,tt2.a,tt2.b 
         FROM t1  
         LEFT OUTER JOIN t2 AS tt2 ON t1.a = tt2.a WHERE tt2.b > 20;
                             QUERY PLAN                              
---------------------------------------------------------------------
 Hash Join  (cost=47.66..86.21 rows=753 width=16)
   Hash Cond: (t1.a = tt2.a)
   ->  Seq Scan on t1  (cost=0.00..32.60 rows=2260 width=8)
   ->  Hash  (cost=38.25..38.25 rows=753 width=8)
         ->  Seq Scan on t2 tt2  (cost=0.00..38.25 rows=753 width=8)
               Filter: (b > 20)
(6 rows)
```

To my understanding (/cc @SStorm) the workaround (moving the 2nd part from the JOIN condition to the WHERE clause) is not valid, as it changes the semantics of the query.

**Comment by matriv** [comment]
> ```sql
> SELECT 
>     c.name
> from core.clusters c
>     JOIN core.projects p ON c.project_id = p.id
>     JOIN core.organizations o ON p.organization_id = o.id
>     LEFT JOIN core.audit_logs al2 ON (c.id = al2.context['cluster']['id'])
>     LEFT JOIN core.stripe_customers sc ON sc.organization_id = o.id
> where c.is_deleted = false
> ```

I have never managed to receive 0 rows with this one (on the cluster in question, and also locally)

**Comment by matriv** [comment]
> ```sql
> SELECT 
>     c.name
> from core.clusters c
>     JOIN core.projects p ON c.project_id = p.id
>     JOIN core.organizations o ON p.organization_id = o.id
>     LEFT JOIN core.audit_logs al2 ON (c.id = al2.context['cluster']['id'])
>     LEFT JOIN core.stripe_customers sc ON sc.organization_id = o.id
> where c.is_deleted = false
> ```
> 
> is enough to sometimes get `0` results
> 
> the `context` field is of type `OBJECT(IGNORED)` which might also have an effect on the physical plan?!
> 
> ```sql
> CREATE TABLE  "core"."audit_logs"
> (
>     "id"                   TEXT NOT NULL,
>     "action"               TEXT,
>     "context"              OBJECT(IGNORED),
>     PRIMARY KEY ("id")
> ) CLUSTERED BY ("id")
> ```

@proddata I have never managed to receive 0 rows with this one (on the cluster in question, and also locally)

**Comment by proddata** [comment]
>  I have never managed to receive 0 rows with this one (on the cluster in question, and also locally)

<img width="710" alt="image" src="https://user-images.githubusercontent.com/23557193/210536875-4c861412-2a0c-4525-8dfe-5865cef8c658.png">

less common, but it still happens

**Comment by matriv** [comment]
@proddata Indeed, now it's easily reproducable with this query as well.

**Comment by matriv** [comment]
Here you can find an Integration test (along with changes to force flipping to RIGHT join) that reproduces the issue: https://github.com/crate/crate/compare/mt/test-flaky?expand=1

**Comment by seut** [comment]
@SStorm Thank you reporting this issue. It was a bit hard to track it down to the root cause and such took a bit longer to fix. It's now fixed it by #13780 and will be released with the next upcoming hotfix release.
